fix(ml): retry OCR OrtSession with remaining providers

apetersson opened the pull request:
## Description

  - Extend the OCR detection and recognition pipelines so that when the leading ONNX Runtime provider (e.g., CoreML) throws an ONNXRuntimeError, we drop just that provider, rebuild the OrtSession with the remaining providers in the existing preference order, and retry the inference. This keeps OCR tasks alive on CPUs while still attempting faster providers first.
  - Add regression coverage to prove the behavior by stubbing RapidOCR to fail on the first call and verifying we retry with ["CPUExecutionProvider"] and still return results.

  Fixes #23391
https://github.com/immich-app/immich/issues/23391

  ## How Has This Been Tested?

  My workstation (Apple Silicon, macOS 15):

  - [x] UV_CACHE_DIR=.uv-cache UV_HTTP_TIMEOUT=120 uv run --extra cpu --group dev pytest test_main.py::TestOcrFallback -q
  <details><summary><h2>Screenshots (if appropriate)</h2></summary>
  <!-- Images go below this line. -->
  </details>

  ## Checklist:

  - [x] I have performed a self-review of my own code
  - [ ] I have made corresponding changes to the documentation if applicable
  - [x] I have no unrelated changes in the PR.
  - [x] I have confirmed that any new dependencies are strictly necessary.
  - [x] I have written tests for new code (if applicable)
  - [x] I have followed naming conventions/patterns in the surrounding code
  - [ ] All code in src/services/ uses repositories implementations for database calls, filesystem operations, etc.
  - [ ] All code in src/repositories/ is pretty basic/simple and does not have any immich specific logic (that belongs in src/services/)

  ## Please describe to which degree, if any, an LLM was used in creating this pull request.

  LLM (ChatGPT/GPT-5) assisted with brainstorming the fallback approach and drafting this description; all code and tests were written and validated hybrid with codex and by hand..

apetersson committed:
fix(ml): retry OCR OrtSession with remaining providers

  Wrap the RapidOCR detection/recognition runs so an ONNXRuntimeError causes us  to drop the failing provider, rebuild the OrtSession, and continue with the  remaining providers in preference order (e.g. CoreML → CPU). This preserves OCR  service when CoreML rejects P-OCRv5 tensors and keeps other providers
  available if they work.

apetersson committed:
fix(ml): test for retry OCR OrtSession with remaining providers

add test_main.py::TestOcrFallback

apetersson commented:
updated my local branch after a small cleanup

mertalev reviewed (commented):
I think it'd be better to have a general fallback to only CPUExecutionProvider if loading the model fails with another provider. It shouldn't be specific to OCR.

apetersson committed:
fix(machine-learning): stabilize ORT fallback across models

  - add provider guard for CLIP encoders to skip CoreML-only setups and log the chosen fallback
  - rebuild face/ocr wrappers after CPU fallback and document why the logic isn’t confined to OrtSession
  - extend fallback unit coverage for face recognition and CLIP provider selection

Tests: uv run pytest test_main.py::TestOrtSession::test_falls_back_to_cpu_when_session_creation_fails
  test_main.py::TestFaceRecognition::test_detection_fallback_retries_with_cpu_provider
  test_main.py::TestFaceRecognition::test_recognition_fallback_retries_with_cpu_provider test_main.py::TestCLIP::test_visual_encoder_prefers_cpu_when_only_coreml
  test_main.py::TestCLIP::test_textual_encoder_prefers_cpu_when_only_coreml

apetersson committed:
fix(ml): clarify coreml fallback logging and mock providers in tests

apetersson committed:
fix(ml): pass mypy with explicit NumPy arrays

mertalev commented:
The facial recognition and OCR code is going through significant changes that mean we no longer depend on libraries for the core inference. This PR's implementation bleeds ORT details in a lot of places where it should be a small self-contained change to fall back to CPU in OrtSession, so I think it's best to close it as stale.
